docs/architecture/adr-050-auto-compound-rewards.md

```diff
@@ -0,0 +1,167 @@
+this, where a simple method would to introduce a new message type, `MsgEnableAutoCompoundRewards`,
+defined as follows:
+
+```protobuf
+message MsgEnableAutoCompoundRewards {
+  option (cosmos.msg.v1.signer) = "delegator_address";
+
+  // ...
+
+  string delegator_address     = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  string src_validator_address = 2 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  string dst_validator_address = 3 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+}
+```
+
+Recall, `x/distribution` executes reward withdrawal via a tuple, (`DelegatorAddress`, `ValidatorAddress`).
+Furthermore, a delegator can also define a different address to which the rewards
+are withdrawn to. This means that we need to know the tuple to execute the withdraw
+and the withdraw address in order to send the rewards from when auto-compounding.
+
+To reflect this in `MsgEnableAutoCompoundRewards`, the `delegator_address` and
+`src_validator_address` fields act as the tuple to execute the withdraw. We can
+use `delegator_address` to find the withdraw address. Finally, the `dst_validator_address`
+defines the validator address to delegate the withdrawn rewards to, from the withdraw address.
+We imagine in most instances, `src_validator_address` and `dst_validator_address`
+will be the same.
+
+When a delegator wants to have their "unrealized" rewards be withdrawn and
+automatically delegated to the relative validator(s), they would broadcast a
+`MsgEnableAutoCompoundRewards` transaction. To stop or disable auto-compounding,
+the user would send a similar transaction as defined by `MsgDisableAutoCompoundRewards`:
+
+```protobuf
+message MsgDisableAutoCompoundRewards {
+  option (cosmos.msg.v1.signer) = "delegator_address";
+
+  // ...
+
+  string delegator_address     = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  string src_validator_address = 2 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+}
+```
+
+In addition, we require the `x/distribution` module to use an additional state
+index to store the records for delegators. When a user submits a `MsgEnableAutoCompoundRewards`
+transaction, we store a record with the following key and value:
+
+```text
+<prefixByte> | delegator_address | src_validator_address -> dst_validator_address
+```
+
+> Note, both `delegator_address` and `src_validator_address` are length-prefixed
+> via `address.MustLengthPrefix` in the key.
+
+When a user decides to disable auto-compounding rewards by sending a `MsgDisableAutoCompoundRewards`
+transaction, we delete the record stored under the above key.
+
+Given that we now have such a key ordering in state, we can iterate over the all
+the relevant records using the dedicated `<prefixByte>` using a prefix `KVStore`,
+which will allow us to withdraw rewards and re-delegate to validators. This can
+be defined as follows in `x/distribution`:
+
+```go
+func (k Keeper) AutoCompoundRewards(ctx sdk.Context) {
```

Comment:
in substrate they don't automate this, a user enables compounding and any delegator or the validator can call compound and it compounds everything for their delegators . I believe this was done due to computational limitations of doing it automatically like this. We won't know which direction is best till we try on a live network but something to consider in alternative design

Reply:
yes, this is why we perform this on epochs. I.e. once every 24h.